File: backend/app/scrapers/fivech.py
"""
5ch（旧2ch）のスクレイピング機能
"""
import httpx
from bs4 import BeautifulSoup
from typing import List, Dict
import logging
import re

logger = logging.getLogger(__name__)

# 5chの主要な板（カテゴリ別）
POPULAR_BOARDS = [
    {"title": "ニュース速報+", "url": "https://asahi.5ch.net/newsplus/", "category": "ニュース"},
    {"title": "ニュース速報", "url": "https://hayabusa9.5ch.net/news/", "category": "ニュース"},
    {"title": "芸スポ速報+", "url": "https://hayabusa9.5ch.net/mnewsplus/", "category": "芸能"},
    {"title": "なんでも実況J", "url": "https://eagle.5ch.net/livejupiter/", "category": "実況"},
    {"title": "プログラマー", "url": "https://mevius.5ch.net/tech/", "category": "PC・技術"},
    {"title": "プログラム", "url": "https://mevius.5ch.net/prog/", "category": "PC・技術"},
]

async def fetch_5ch_boards() -> List[Dict]:
    """5chの板一覧を取得（人気板のリストを返す）"""
    # 簡易実装：予め定義された人気板のリストを返す
    # 完全な板一覧の取得は複雑なため、主要な板のみ提供
    return POPULAR_BOARDS

async def fetch_5ch_threads(board_url: str, limit: int = 50) -> List[Dict]:
    """指定された板のスレッド一覧を取得"""
    try:
        # subject.txt を取得（5chの標準的なスレッド一覧フォーマット）
        subject_url = board_url.rstrip('/') + '/subject.txt'

        async with httpx.AsyncClient(timeout=30.0) as client:
            # 5chサーバーへのリクエストには適切なUser-Agentが必要
            headers = {
                'User-Agent': 'Mozilla/5.0 (compatible; EsunaBot/0.1)',
            }
            response = await client.get(subject_url, headers=headers)
            response.encoding = 'cp932'  # 5chの文字コード
            response.raise_for_status()

            return parse_5ch_threads(response.text, board_url, limit)
    except Exception as e:
        logger.error(f"Error fetching 5ch threads from {board_url}: {e}")
        raise

def parse_5ch_threads(subject_txt: str, board_url: str, limit: int) -> List[Dict]:
    """subject.txt をパースしてスレッド一覧を作成"""
    threads = []
    lines = subject_txt.strip().split('\n')[:limit]

    for line in lines:
        try:
            # subject.txt のフォーマット: "datファイル名.dat<>スレッドタイトル (レス数)"
            # 例: "1234567890.dat<>テストスレッド (100)"
            if '<>' not in line:
                continue

            dat_file, title_with_count = line.split('<>', 1)
            thread_id = dat_file.replace('.dat', '')

            # レス数を抽出
            response_count = 0
            match = re.search(r'\((\d+)\)', title_with_count)
            if match:
                response_count = int(match.group(1))
                title = re.sub(r'\s*\(\d+\)\s*$', '', title_with_count)
            else:
                title = title_with_count

            # スレッドURLを構築
            thread_url = f"{board_url.rstrip('/')}/test/read.cgi/{thread_id}/"

            threads.append({
                "title": title,
                "url": thread_url,
                "response_count": response_count,
                "thread_id": thread_id
            })
        except Exception as e:
            logger.warning(f"Error parsing thread line: {line} - {e}")
            continue

    return threads

async def fetch_5ch_posts(thread_url: str, start: int = 1, end: int = 100) -> List[Dict]:
    """スレッドの投稿を取得"""
    try:
        # dat形式で取得（より軽量）
        # URLからthread_idを抽出
        match = re.search(r'/(\d+)/?$', thread_url.rstrip('/'))
        if not match:
            raise ValueError(f"Invalid thread URL: {thread_url}")

        thread_id = match.group(1)
        board_url = re.sub(r'/test/read\.cgi/\d+/?$', '', thread_url)
        dat_url = f"{board_url}/{thread_id}.dat"

        async with httpx.AsyncClient(timeout=30.0) as client:
            headers = {
                'User-Agent': 'Mozilla/5.0 (compatible; EsunaBot/0.1)',
            }
            response = await client.get(dat_url, headers=headers)
            response.encoding = 'cp932'
            response.raise_for_status()

            return parse_5ch_posts(response.text, start, end)
    except Exception as e:
        logger.error(f"Error fetching 5ch posts from {thread_url}: {e}")
        raise

def parse_5ch_posts(dat_content: str, start: int, end: int) -> List[Dict]:
    """dat形式の投稿をパース"""
    posts = []
    lines = dat_content.strip().split('\n')

    for i, line in enumerate(lines, start=1):
        if i < start or i > end:
            continue

        try:
            # dat形式: "名前<>メール<>日付 ID<>本文<>スレタイ"
            parts = line.split('<>')
            if len(parts) < 4:
                continue

            name = parts[0]
            mail = parts[1]
            datetime_id = parts[2]
            text = parts[3]

            # HTMLタグを除去
            text = re.sub(r'<br>', '\n', text)
            text = re.sub(r'<[^>]+>', '', text)

            posts.append({
                "number": i,
                "name": name,
                "datetime": datetime_id,
                "text": text,
                "mail": mail
            })
        except Exception as e:
            logger.warning(f"Error parsing post line {i}: {e}")
            continue

    return posts

File: backend/src/scrapers/fivech.ts
/**
 * 5ch（旧2ch）のスクレイピング機能
 */
import type { FivechBoard, FivechThread, FivechPost } from "../types";

/** 5chの主要な板（カテゴリ別） */
const POPULAR_BOARDS: FivechBoard[] = [
  {
    title: "ニュース速報+",
    url: "https://asahi.5ch.net/newsplus/",
    category: "ニュース",
  },
  {
    title: "ニュース速報",
    url: "https://hayabusa9.5ch.net/news/",
    category: "ニュース",
  },
  {
    title: "芸スポ速報+",
    url: "https://hayabusa9.5ch.net/mnewsplus/",
    category: "芸能",
  },
  {
    title: "なんでも実況J",
    url: "https://eagle.5ch.net/livejupiter/",
    category: "実況",
  },
  {
    title: "プログラマー",
    url: "https://mevius.5ch.net/tech/",
    category: "PC・技術",
  },
  {
    title: "プログラム",
    url: "https://mevius.5ch.net/prog/",
    category: "PC・技術",
  },
];

/**
 * 5chの板一覧を取得（人気板のリストを返す）
 */
export async function fetch5chBoards(): Promise<FivechBoard[]> {
  return POPULAR_BOARDS;
}

/**
 * subject.txt をパースしてスレッド一覧を作成
 */
function parse5chThreads(
  subjectTxt: string,
  boardUrl: string,
  limit: number
): FivechThread[] {
  const threads: FivechThread[] = [];
  const lines = subjectTxt.trim().split("\n").slice(0, limit);

  for (const line of lines) {
    try {
      if (!line.includes("<>")) continue;

      const [datFile, titleWithCount] = line.split("<>", 2);
      const threadId = datFile.replace(".dat", "");

      // レス数を抽出
      let responseCount = 0;
      let title = titleWithCount;
      const match = titleWithCount.match(/\((\d+)\)/);
      if (match) {
        responseCount = parseInt(match[1], 10);
        title = titleWithCount.replace(/\s*\(\d+\)\s*$/, "");
      }

      // スレッドURLを構築
      const parsed = new URL(boardUrl);
      const boardName = parsed.pathname.replace(/^\/|\/$/g, "");
      const threadUrl = `${parsed.protocol}//${parsed.host}/test/read.cgi/${boardName}/${threadId}/`;

      threads.push({
        title,
        url: threadUrl,
        response_count: responseCount,
        thread_id: threadId,
      });
    } catch (e) {
      console.warn(`Error parsing thread line: ${line}`, e);
    }
  }

  return threads;
}

/**
 * 指定された板のスレッド一覧を取得
 */
export async function fetch5chThreads(
  boardUrl: string,
  limit: number = 50
): Promise<FivechThread[]> {
  const subjectUrl = boardUrl.replace(/\/$/, "") + "/subject.txt";

  const response = await fetch(subjectUrl, {
    headers: {
      "User-Agent": "Mozilla/5.0 (compatible; EsunaBot/0.2)",
    },
    signal: AbortSignal.timeout(30_000),
  });
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
  }

  // 5chはShift_JIS（cp932）
  const buffer = await response.arrayBuffer();
  const decoder = new TextDecoder("shift_jis");
  const text = decoder.decode(buffer);

  return parse5chThreads(text, boardUrl, limit);
}

/**
 * dat形式の投稿をパース
 */
function parse5chPosts(
  datContent: string,
  start: number,
  end: number
): FivechPost[] {
  const posts: FivechPost[] = [];
  const lines = datContent.trim().split("\n");

  for (let i = 0; i < lines.length; i++) {
    const lineNum = i + 1;
    if (lineNum < start || lineNum > end) continue;

    try {
      const parts = lines[i].split("<>");
      if (parts.length < 4) continue;

      const name = parts[0];
      const mail = parts[1];
      const datetimeId = parts[2];
      let text = parts[3];

      // HTMLタグを除去
      text = text.replace(/<br>/gi, "\n");
      text = text.replace(/<[^>]+>/g, "");

      posts.push({
        number: lineNum,
        name,
        datetime: datetimeId,
        text,
        mail,
      });
    } catch (e) {
      console.warn(`Error parsing post line ${lineNum}:`, e);
    }
  }

  return posts;
}

/**
 * スレッドの投稿を取得
 */
export async function fetch5chPosts(
  threadUrl: string,
  start: number = 1,
  end: number = 100
): Promise<FivechPost[]> {
  // thread_url 例: https://asahi.5ch.net/test/read.cgi/newsplus/1234567890/
  // dat URL 例:    https://asahi.5ch.net/newsplus/dat/1234567890.dat
  const match = threadUrl.match(/\/test\/read\.cgi\/([^/]+)\/(\d+)\/?$/);
  if (!match) {
    throw new Error(`Invalid thread URL: ${threadUrl}`);
  }

  const boardName = match[1];
  const threadId = match[2];
  const parsed = new URL(threadUrl);
  const datUrl = `${parsed.protocol}//${parsed.host}/${boardName}/dat/${threadId}.dat`;

  const response = await fetch(datUrl, {
    headers: {
      "User-Agent": "Mozilla/5.0 (compatible; EsunaBot/0.2)",
    },
    signal: AbortSignal.timeout(30_000),
  });
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
  }

  // 5chはShift_JIS（cp932）
  const buffer = await response.arrayBuffer();
  const decoder = new TextDecoder("shift_jis");
  const text = decoder.decode(buffer);

  return parse5chPosts(text, start, end);
}
